runtime/libffi/z/ffi64.c

```diff
@@ -875,58 +921,64 @@ ffi_prep_cif_machdep(ffi_cif *cif)
 	{
 		type = ffi_check_struct_type (*ptr);
 
+      (*ptr)->type = ffi_check_struct_for_complex(*ptr);
+
+
 	  /* If we pass the struct via pointer, we must reserve space
 	     to copy its data for proper call-by-value semantics.  */
 	  if (type == FFI_TYPE_POINTER)
 	    struct_size += ROUND_SIZE ((*ptr)->size);
 	}
 
       /* Now handle all primitive int/float data types.  */
-      switch (type)
+    switch (type)
       {
-       	/* The first MAX_FPRARGS floating point arguments
-	     go in FPRs, the rest overflow to the stack.  */
-
-	case FFI_TYPE_LONGDOUBLE:
-	  if (n_fpr < MAX_FPRARGS)
-	    n_fpr+=2;
-	  else
-	    n_ov += sizeof(long double) / sizeof(long);
-	  break;
-
-	case FFI_TYPE_COMPLEX:
-       	case FFI_TYPE_DOUBLE:
-       	case FFI_TYPE_FLOAT:
-	  if (n_fpr < MAX_FPRARGS)
-	    n_fpr++;
-	  else
-	    n_ov += sizeof (double) / sizeof (long);
-	  break;
-
-	  /* On 31-bit machines, 64-bit integers are passed in GPR pairs,
-	     if one is still available, or else on the stack.  If only one
-	     register is free, skip the register (it won't be used for any
-	     subsequent argument either).  */
-
-	case FFI_TYPE_UINT64:
-       	case FFI_TYPE_SINT64:
-	  if (n_gpr == MAX_GPRARGS-1)
-	    n_gpr = MAX_GPRARGS;
-	  if (n_gpr < MAX_GPRARGS)
-	    n_gpr += 2;
-	  else
-	    n_ov += 2;
-	  break;
+      /**
+       * The first MAX_FPRARGS floating point arguments
+       * go in FPRs, the rest overflow to the stack.
+       */
```

Comment:
Please revert all the cosmetic changes to this function. The only significant changes are the additions of lines 910 and 924. The cosmetic changes should be addressed in a separate pull request.